spec/framework/core/fact_manager_spec.rb

```diff
@@ -4,90 +4,349 @@
+        end
 
-      allow(cache_manager)
-        .to receive(:resolve_facts)
-        .with(seached_facts)
-        .and_return([seached_facts, []])
+        it 'does not load core and external facts' do
+          Facter::FactManager.instance.resolve_fact(user_query)
 
-      allow(cache_manager)
-        .to receive(:cache_facts)
-        .with([resolved_fact])
+          expect(logger).not_to have_received(:debug)
+            .with("Searching fact: #{user_query} in core facts and external facts")
+        end
 
-      resolved_facts = Facter::FactManager.instance.resolve_facts(user_query)
+        it 'does not load all custom facts' do
+          Facter::FactManager.instance.resolve_fact(user_query)
 
-      expect(resolved_facts).to eq([resolved_fact])
+          expect(logger).not_to have_received(:debug)
+            .with("Searching fact: #{user_query} in all custom facts")
+        end
+
+        it 'resolves fact' do
+          resolved_facts = Facter::FactManager.instance.resolve_fact(user_query)
+
+          expect(resolved_facts).to eql([resolved_fact])
+        end
+      end
+
+      context 'when is not found in custom_dir/fact_name.rb' do
+        before do
+          # mock custom_fact_by_filename to return nil
+          allow(Facter::FactLoader.instance).to receive(:load_custom_fact).and_return([])
+          stub_query_parser([[user_query], []], [])
+          stub_external_manager(searched_facts, [])
+          stub_cache_manager([], [])
+
+          # mock core_or_external_fact to return nil
+          allow(Facter::FactLoader.instance).to receive(:load_internal_facts).and_return([])
+          stub_query_parser([[user_query], []], [])
+          stub_internal_manager([], [])
+          stub_external_manager([], [])
+          stub_cache_manager([], [])
+
+          # mock all_custom_facts to return resolved_fact
+          allow(Facter::FactLoader.instance).to receive(:load_custom_facts).and_return(loaded_facts)
+          stub_query_parser([[user_query], loaded_facts], searched_facts)
+          stub_external_manager(searched_facts, [resolved_fact])
+          stub_cache_manager(searched_facts, [])
+        end
+
+        it 'tries to load it from fact_name.rb' do
+          Facter::FactManager.instance.resolve_fact(user_query)
+
+          expect(logger).to have_received(:debug)
+            .with("Searching fact: #{user_query} in file: #{user_query}.rb")
+        end
+
+        it 'loads core and external facts' do
+          Facter::FactManager.instance.resolve_fact(user_query)
+
+          expect(logger).to have_received(:debug)
+            .with("Searching fact: #{user_query} in core facts and external facts")
+        end
+
+        it 'loads all custom facts' do
+          Facter::FactManager.instance.resolve_fact(user_query)
+
+          expect(logger).to have_received(:debug)
+            .with("Searching fact: #{user_query} in all custom facts")
+        end
```

Comment:
Do we want to maybe assert that this behavior happens in a specific order?

Reply:
Tried, but it will break rubocop multiple expectations cops. On the other hand, given the way tests are executed, if the order is not expected, and something else gets on the logger, the expectation will fail.